src/components/code-editor/index.js

```diff
@@ -104,53 +115,72 @@ module.exports = {
 		    modes = ["yaml"];
 		    mimeType = "text/x-yaml";
 	    }
-	    var readOnly = ! this.file.isWritable();
+	    var readOnly = ! this.currentFile.isWritable();
 
-	    this.file.getInputStream(this.context.network, this.context.crypto, props.sizeHigh(), props.sizeLow(), function(read){})
+	    this.currentFile.getInputStream(this.context.network, this.context.crypto, props.sizeHigh(), props.sizeLow(), function(read){})
 	        .thenCompose(function(reader) {
                 var size = that.getFileSize(props);
                 var data = convertToByteArray(new Int8Array(size));
                 return reader.readIntoArray(data, 0, data.length)
                     .thenApply(function(read){
-                        iframe.contentWindow.postMessage({modes:modes, mime:mimeType, readOnly:readOnly, text:new TextDecoder().decode(data)}, '*');
+                        setTimeout(function(){
+                            iframe.contentWindow.postMessage({modes:modes, mime:mimeType, readOnly:readOnly, text:new TextDecoder().decode(data)}, '*');
+                        });
                     });
 	    });
 	},
 
 	getAndSave: function() {
+	    if(this.saving) {
+	        return;
+	    }
 	    var iframe = document.getElementById("editor");
 	    this.expectingSave = true;
 	    iframe.contentWindow.postMessage({type:"save"}, '*');
 	},
 
     save: function(text) {
+	    this.saving = true;
 	    var bytes = convertToByteArray(new TextEncoder().encode(text));
 	    var java_reader = peergos.shared.user.fs.AsyncReader.build(bytes);
-
-	    const file = this.file;
+	    const file = this.currentFile;
 	    const context = this.context;
 	    const size = text.length;
-	    const parent = this.parent;
 	    const that = this;
-	    this.saving = true;
-	    this.parent.uploadFileJS(file.getName(), java_reader, (size - (size % Math.pow(2, 32)))/Math.pow(2, 32), size,
-				true, true, context.network, context.crypto, len => {}, context.getTransactionService())
-        .thenApply(function(updated) {
+        this.currentParent.uploadFileJS(file.getName(), java_reader, (size - (size % Math.pow(2, 32)))/Math.pow(2, 32), size,
```

Comment:
this is where we should use a method on the file itself, rather than the parent

Reply:
This is also not doing a CAS

Reply:
When i run a test highlighting a conncurrent modification the upload fails with a CAS exception in setPointer. if i try again it succeeds - no idea why.  i'm going to need your assistance.

Reply:
We need a new method on fileWrapper - overwriteSection(long offset, Snapshot expectedVersion, data ...) which uses the cas of the existing version chunks in the champ